// dapple-aws/index.js
// Generated by CoffeeScript 1.6.3
var AWS, DappleAws, fs;

fs = require('fs');

AWS = require('aws-sdk');

AWS.config.update({
  accessKeyId: process.env['AWS_KEY'],
  secretAccessKey: process.env['AWS_SECRET'],
  region: 'us-east-1'
});

DappleAws = (function() {
  function DappleAws(bucketName) {
    this.bucketName = bucketName;
  }

  DappleAws.prototype.awsUrl = function(filename) {
    return "https://s3.amazonaws.com/" + this.bucketName + "/" + filename;
  };

  DappleAws.prototype.filePathToName = function(filePath) {
    return filePath.match(/[^\/]*$/)[0];
  };

  DappleAws.prototype.upload = function(filePath, callback) {
    var self;
    self = this;
    return fs.readFile(filePath, function(err, data) {
      var filename, s3;
      if (err) {
        console.log("AWS Error Reading File: ", err);
        callback(err, null);
        return;
      }
      filename = self.filePathToName(filePath);
      s3 = new AWS.S3({
        params: {
          Bucket: self.bucketName,
          Key: filename
        }
      });
      return s3.putObject({
        Body: data
      }, function(err, data) {
        if (err) {
          console.log("AWS Error Uploading File: ", err);
          callback(err, null);
          return;
        }
        return callback(null, self.awsUrl(filename));
      });
    });
  };

  return DappleAws;

})();

module.exports = DappleAws;
